Use tini to prevent JVM as pid1 errors

https://github.com/krallin/tini

quick summary is running JVMs as pid1 in dockerfiles cause problems -- search for `Caused by: com.sun.tools.attach.AttachNotSupportedException: Unable to get pid of LinuxThreads manager thread at sun.tools.attach.LinuxVirtualMachine.<init>(LinuxVirtualMachine.java:86)` to see examples

The full stacktrace of what we were seeing is available here: https://github.com/palantir/http-remoting/issues/403

Here's some more analysis I did internally on the stacktrace:

> ```
> Caused by: com.sun.tools.attach.AttachNotSupportedException: Unable to get pid of LinuxThreads manager thread
>     at sun.tools.attach.LinuxVirtualMachine.<init>(LinuxVirtualMachine.java:86)
>     ... 64 more
> ```
> 
> The reason PID1 is important is that there's a bug in the JRE that doesn't correctly handle the JRE being pid1 (and having more child processes than expected).
> 
> The above stacktrace has [LinuxVirtualMachine.java/#86](http://grepcode.com/file/repository.grepcode.com/java/root/jdk/openjdk/8u40-b25/sun/tools/attach/LinuxVirtualMachine.java/#86) in it from the call to native method `getLinuxThreadsManager` a few lines earlier.  [LinuxVirtualMachine.c#255](http://code.metager.de/source/xref/openjdk/jdk8/jdk/src/solaris/native/sun/tools/attach/LinuxVirtualMachine.c#255)
> 
> The purpose of that method seems to be to get the main linux thread from which other linux threads of the JVM were spawned from.
> 
> It does this by counting how many children the current pid has, by looking through all files that match `/proc/[0-9]+` in [forEachProcess](http://code.metager.de/source/xref/openjdk/jdk8/jdk/src/solaris/native/sun/tools/attach/LinuxVirtualMachine.c#forEachProcess)
> 
> Of course because the JVM is pid 1, every other process in the docker image has pid 1 as its parent.  So the count of processes with ppid==1 is > 1.
> 
> So the `if` statements at [LinuxVirtualMachine.c#273](http://code.metager.de/source/xref/openjdk/jdk8/jdk/src/solaris/native/sun/tools/attach/LinuxVirtualMachine.c#273) and [LinuxVirtualMachine.c#285](http://code.metager.de/source/xref/openjdk/jdk8/jdk/src/solaris/native/sun/tools/attach/LinuxVirtualMachine.c#285) don't hit, and the C function throws the exception, which is then seen in the stacktrace:
> 
> Any method that eventually calls `sun.tools.attach.LinuxVirtualMachine.<init>` triggers this code path and causes an exception, and the ALPN injection calls this as well as some other JVM cli tools like `jmap` and `jstack`.

--- a/resource-managers/kubernetes/docker-minimal-bundle/src/main/docker/driver/Dockerfile
+++ b/resource-managers/kubernetes/docker-minimal-bundle/src/main/docker/driver/Dockerfile
@@ -22,7 +22,7 @@ FROM openjdk:8-alpine
 # docker build -t spark-driver:latest -f dockerfiles/driver/Dockerfile .
 
 RUN apk upgrade --update
-RUN apk add --update bash
+RUN apk add --update bash tini
 RUN mkdir -p /opt/spark
 RUN touch /opt/spark/RELEASE
 
@@ -41,4 +41,4 @@ CMD SPARK_CLASSPATH="${SPARK_HOME}/jars/*" && \
     if ! [ -z ${SPARK_SUBMIT_EXTRA_CLASSPATH+x} ]; then SPARK_CLASSPATH="$SPARK_SUBMIT_EXTRA_CLASSPATH:$SPARK_CLASSPATH"; fi && \
     if ! [ -z ${SPARK_EXTRA_CLASSPATH+x} ]; then SPARK_CLASSPATH="$SPARK_EXTRA_CLASSPATH:$SPARK_CLASSPATH"; fi && \
     if ! [ -z ${SPARK_MOUNTED_FILES_DIR} ]; then cp -R "$SPARK_MOUNTED_FILES_DIR/." .; fi && \
-    exec ${JAVA_HOME}/bin/java $SPARK_DRIVER_JAVA_OPTS -cp $SPARK_CLASSPATH -Xms$SPARK_DRIVER_MEMORY -Xmx$SPARK_DRIVER_MEMORY $SPARK_DRIVER_CLASS $SPARK_DRIVER_ARGS
+    exec /sbin/tini -- ${JAVA_HOME}/bin/java $SPARK_DRIVER_JAVA_OPTS -cp $SPARK_CLASSPATH -Xms$SPARK_DRIVER_MEMORY -Xmx$SPARK_DRIVER_MEMORY $SPARK_DRIVER_CLASS $SPARK_DRIVER_ARGS

--- a/resource-managers/kubernetes/docker-minimal-bundle/src/main/docker/executor/Dockerfile
+++ b/resource-managers/kubernetes/docker-minimal-bundle/src/main/docker/executor/Dockerfile
@@ -22,7 +22,7 @@ FROM openjdk:8-alpine
 # docker build -t spark-executor:latest -f dockerfiles/executor/Dockerfile .
 
 RUN apk upgrade --update
-RUN apk add --update bash
+RUN apk add --update bash tini
 RUN mkdir -p /opt/spark
 RUN touch /opt/spark/RELEASE
 
@@ -41,4 +41,4 @@ CMD SPARK_CLASSPATH="${SPARK_HOME}/jars/*" && \
     if ! [ -z ${SPARK_MOUNTED_CLASSPATH}+x} ]; then SPARK_CLASSPATH="$SPARK_MOUNTED_CLASSPATH:$SPARK_CLASSPATH"; fi && \
     if ! [ -z ${SPARK_EXECUTOR_EXTRA_CLASSPATH+x} ]; then SPARK_CLASSPATH="$SPARK_EXECUTOR_EXTRA_CLASSPATH:$SPARK_CLASSPATH"; fi && \
     if ! [ -z ${SPARK_MOUNTED_FILES_DIR} ]; then cp -R "$SPARK_MOUNTED_FILES_DIR/." .; fi && \
-    exec ${JAVA_HOME}/bin/java -Dspark.executor.port=$SPARK_EXECUTOR_PORT -Xms$SPARK_EXECUTOR_MEMORY -Xmx$SPARK_EXECUTOR_MEMORY -cp $SPARK_CLASSPATH org.apache.spark.executor.CoarseGrainedExecutorBackend --driver-url $SPARK_DRIVER_URL --executor-id $SPARK_EXECUTOR_ID --cores $SPARK_EXECUTOR_CORES --app-id $SPARK_APPLICATION_ID --hostname $SPARK_EXECUTOR_POD_IP
+    exec /sbin/tini -- ${JAVA_HOME}/bin/java -Dspark.executor.port=$SPARK_EXECUTOR_PORT -Xms$SPARK_EXECUTOR_MEMORY -Xmx$SPARK_EXECUTOR_MEMORY -cp $SPARK_CLASSPATH org.apache.spark.executor.CoarseGrainedExecutorBackend --driver-url $SPARK_DRIVER_URL --executor-id $SPARK_EXECUTOR_ID --cores $SPARK_EXECUTOR_CORES --app-id $SPARK_APPLICATION_ID --hostname $SPARK_EXECUTOR_POD_IP

--- a/resource-managers/kubernetes/docker-minimal-bundle/src/main/docker/init-container/Dockerfile
+++ b/resource-managers/kubernetes/docker-minimal-bundle/src/main/docker/init-container/Dockerfile
@@ -22,7 +22,7 @@ FROM openjdk:8-alpine
 # docker build -t spark-executor:latest -f dockerfiles/executor/Dockerfile .
 
 RUN apk upgrade --update
-RUN apk add --update bash
+RUN apk add --update bash tini
 RUN mkdir -p /opt/spark
 RUN touch /opt/spark/RELEASE
 
@@ -35,4 +35,4 @@ ENV SPARK_HOME /opt/spark
 
 WORKDIR /opt/spark
 
-ENTRYPOINT [ "bin/spark-class", "org.apache.spark.deploy.rest.kubernetes.KubernetesSparkDependencyDownloadInitContainer" ]
+ENTRYPOINT [ "/sbin/tini", "--", "bin/spark-class", "org.apache.spark.deploy.rest.kubernetes.KubernetesSparkDependencyDownloadInitContainer" ]

--- a/resource-managers/kubernetes/docker-minimal-bundle/src/main/docker/resource-staging-server/Dockerfile
+++ b/resource-managers/kubernetes/docker-minimal-bundle/src/main/docker/resource-staging-server/Dockerfile
@@ -22,7 +22,7 @@ FROM openjdk:8-alpine
 # docker build -t spark-executor:latest -f dockerfiles/executor/Dockerfile .
 
 RUN apk upgrade --update
-RUN apk add --update bash
+RUN apk add --update bash tini
 RUN mkdir -p /opt/spark
 RUN touch /opt/spark/RELEASE
 
@@ -35,4 +35,4 @@ ENV SPARK_HOME /opt/spark
 
 WORKDIR /opt/spark
 
-ENTRYPOINT [ "bin/spark-class", "org.apache.spark.deploy.rest.kubernetes.ResourceStagingServer" ]
+ENTRYPOINT [ "/sbin/tini", "--", "bin/spark-class", "org.apache.spark.deploy.rest.kubernetes.ResourceStagingServer" ]

--- a/resource-managers/kubernetes/docker-minimal-bundle/src/main/docker/shuffle-service/Dockerfile
+++ b/resource-managers/kubernetes/docker-minimal-bundle/src/main/docker/shuffle-service/Dockerfile
@@ -22,7 +22,7 @@ FROM openjdk:8-alpine
 # docker build -t spark-shuffle:latest -f dockerfiles/shuffle/Dockerfile .
 
 RUN apk upgrade --update
-RUN apk add --update bash
+RUN apk add --update bash tini
 RUN mkdir -p /opt/spark
 RUN touch /opt/spark/RELEASE
 
@@ -36,4 +36,4 @@ ENV SPARK_HOME /opt/spark
 
 WORKDIR /opt/spark
 
-CMD ["/bin/sh","-c","/opt/spark/bin/spark-class org.apache.spark.deploy.ExternalShuffleService 1"]
+ENTRYPOINT [ "/sbin/tini", "--", "bin/spark-class", "org.apache.spark.deploy.ExternalShuffleService", "1" ]

